### PR TITLE
Add bootstrap manager.

### PR DESCRIPTION
**Pull request by piotrm50**
# Description of change

Add bootstrap manager that keeps track of bootstrap status of tangle and notarization manager and triggers an event when both are bootstrapped.

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [ ] My code follows the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/tangle/timemanager.go [hunk 1]
```diff
@@ -189,6 +189,7 @@ func (t *TimeManager) updateSyncedState() {
 		t.Events.SyncChanged.Trigger(&SyncChangedEvent{Synced: newSynced})
 		if newSynced {
 			t.bootstrapped = true
+			t.Events.Bootstrapped.Trigger(&SyncChangedEvent{Synced: true})
```
Comment (karimodm) [inline packages/tangle/timemanager.go:192]: This should be its own event type, without fields: as it can only turn `true`.

File: packages/notarization/manager.go [hunk 2]
```diff
@@ -566,12 +574,22 @@ func (m *Manager) moveLatestCommittableEpoch(currentEpoch epoch.Index) ([]*Epoch
...
 	}
 	for _, manaVectorUpdateEvent := range manaVectorUpdateEvents {
 		m.Events.ManaVectorUpdate.Trigger(manaVectorUpdateEvent)
 	}
 }
 
+func (m *Manager) updateEpochsBootstrapped(ei epoch.Index) {
+	m.epochCommitmentFactoryMutex.Lock()
+	defer m.epochCommitmentFactoryMutex.Unlock()
+	if !m.bootstrapped && ei > epoch.IndexFromTime(clock.SyncedTime().Add(-2*m.options.MinCommittableEpochAge)) {
```
Comment (karimodm) [inline packages/notarization/manager.go:587]: Let's move this "magic number" in a constant, no need to make it configurable.

File: packages/bootstrapmanager/manager.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+// Bootstrapped returns bool indicating if the node is bootstrapped.
+func (m *Manager) Bootstrapped() bool {
+	m.RLock()
+	defer m.RUnlock()
+	return m.tangle.Bootstrapped() && m.notarizationManager.Bootstrapped()
+}
+
+// Events returns the events of the manager.
+func (m *Manager) Events() *Events {
```
Comment (karimodm) [inline packages/bootstrapmanager/manager.go:63]: A while ago we decided not to have getters for Events, but just export the events fields.

**Review by karimodm — DISMISSED** [review]
Awesome! Just a few minor comments.

File: plugins/bootstrapmanager/plugin.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	_          = new(dependencies)
+	pluginDeps = new(pluginDependencies)
+)
+
+type dependencies struct {
+	dig.In
+
+	Tangle          *tangle.Tangle
+	NotarizationMgr *notarization.Manager
+}
```
Comment (jonastheis) [inline plugins/bootstrapmanager/plugin.go:28]: Why are those needed?

**Review by jonastheis — APPROVED** [review]
